Onboarding: Fix plan picker comparison table width on large displays

david-szabo97 opened the pull request:
## Changes proposed in this Pull Request

- Fix comparison table being wider than the container on large desktop screens
- Add padding at the end of the comparison table so that, when scrolling horizontally on narrow screens, it would line up with the rest of the page's content

## Testing instructions

### Prepare for testing

* Check out the branch on your machine and run in two parallel terminal windows:
  - `yarn && yarn start`
  - `cd apps/editing-toolkit && yarn dev --sync`
* Add an _unlaunched_ site created with `/new` to your sandbox (for convenience, `SITE_ID.wordpress.com`)
* Visit the block editor (e.g. `/page/SITE_ID.wordpress.com/home`)
* Click on "Launch" button — Step by Step Launch modal should appear
* Navigate to the Plans step

### Testing checklist

Please carry out the tests **across different browsers** (at least Chrome / Firefox / Safari)

- [x] Check that the left side of the Plan comparison table is always aligned with its container
- [x] Check that, when the viewport is wide enough, the right border of the Plan comparison table is aligned with its container
  - [x] Check that, when the viewport is not wide enough, the table overflows its container to the right (and the section can be scrolled horizontally), while maintaining the right padding


Additionally, **test for potential regressions in other flows** importing `@automattic/onboarding` and `@automattic/plans-grid`:

- [x] New Onboarding
- [x] Focused Launch

| before | after |
| - | - |
| ![image](https://user-images.githubusercontent.com/2256104/101488954-8c64ec00-3960-11eb-839f-99febc91e76e.png) | ![image](https://user-images.githubusercontent.com/2256104/101488736-4871e700-3960-11eb-969e-feec3f980f22.png) |


Fixes #47883

david-szabo97 committed:
Fix width

p-jackson reviewed (commented):
Fix works for me 👍

What do you think about moving the new breakpoint to the `onboarding-block-edge2edge-container` mixin [here](https://github.com/Automattic/wp-calypso/blob/0eb078457ddf4fc756ccc008718f1168fcd1b501/packages/onboarding/styles/mixins.scss#L157)? It's where the other break points are and it seems related.

Something I wanted to note, but I personally think this is fine:
<img width="982" alt="image" src="https://user-images.githubusercontent.com/1500769/101547551-4c107700-3a0f-11eb-894c-c47b9b346fef.png">
When the screen gets small the details grid scrolls horizontally. Good. But the right-margin also disappears, I was expecting that if I scrolled all the way to the right that there'd be a margin so the table lines up. Not a big deal though.

david-szabo97 committed:
Move break to mixins

david-szabo97 committed:
Add padding to the right side when scrolling

david-szabo97 commented:
Thanks @p-jackson ! Moved the breaks to the mixin and fixed the missing padding as well! Feels a lot better!

p-jackson reviewed (commented):
Definitely!

Another wee breakpoint issue: when the browser is smaller than `huge` but not so small that it's horizontally scrolling, the tables don't line up.
<img width="1244" alt="image" src="https://user-images.githubusercontent.com/1500769/101696300-cd810b80-3ada-11eb-89c0-33f5dcd34adb.png">

It happens in the launch flow and in `/new/plans-modal` so I wonder if the issue is in the rules added to the `plans-grid` package since that code's shared.

david-szabo97 commented:
Oof. I can't think of any solution for that in CSS. We can't use a breakpoint here because the table's width may vary. Any idea? I would avoid JS for this.

ciampo commented:
I've updated the PR description with more detailed testing instructions. I'll try to pick this issue up in my rotation if nothing urgent comes up

ciampo committed:
Move padding from container to the plans table itself, and remove unnecessary width rules

ciampo committed:
Use breakpoint variable insead of hardcoded number

ciampo committed:
Fix: set margin and padding to 0 on huge screens

This is mainly a fix for gutenboarding, as otherwise the table was already
working fine in Step by step flow.

ciampo committed:
Fix: media query breakpoint overlap

ciampo commented:
Rebased and pushed a few changes.

Hey @david-szabo97, in order to avoid the issue flagged by @p-jackson (the table not lining up on desktop large) I took a slightly different approach on how to achieve a full bleed, scrolling table. The main ideas of this new approach are:

- moved the padding from the container to the table itself
- removed the `width` calculations, and replaced them with `width: auto` (seems to be enough)
- refactored the code to use the `$plans-grid-max-page-width` in more places. I also changed the value of that variable to be the same as the `huge` breakpoint, and fixed a small bug where there was a `1px` overlap between the `min-width` and `max-width` media queries

I'm not going to be around until the 4th of Jan, but I tried to split my changes across different commits to help the review process. **Please do thoroughly review this changes across different browsers and test Gutenboarding (and Focused Launch) flows too for any regressions.**

Finally, I'm also adding @yansern to the list of reviewers, as he's worked on this part of the codebase before and may have some useful insights.

ciampo commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/plan-step/styles.scss:
```diff
@@ -10,5 +11,10 @@
 		position: relative;
 		@include onboarding-block-edge2edge-container;
+
+		@media ( min-width: $plans-grid-max-page-width ) {
+			margin-left: 0;
+			margin-right: 0;
+		}
```
This change lies in Editing toolkit because it only applies in Step by Step launch (not gutenbording).

ciampo commented on packages/onboarding/styles/mixins.scss:
```diff
@@ -157,15 +157,7 @@
-	@include break-medium {
-		width: calc( 100% + #{$onboarding-block-margin-medium * 2} );
-	}
+	width: auto;
```
`width: auto` seems to be enough to force the element to the correct width, taking into account the negative margins

ciampo commented on packages/plans-grid/src/plans-grid/style.scss:
```diff
@@ -40,13 +25,12 @@
 		position: absolute;
 		left: 0;
-
-		@include plans-grid-edge2edge-padding;
```
Padding has been moved from `.plans-grid__details-container` to `.plans-details__table` (see changes to `packages/plans-grid/src/plans-details/style.scss` in this PR)

yansern commented:
All browsers renders as described in the test description except for Safari Desktop. Also tested on iOS simulator.
I'm rechecking Safari Desktop.

yansern reviewed (approved):
I'm approving this PR. Everything checks out except for Safari as I'm unable to get Safari to read my hosts file. We can roll out another PR for Safari if its an issue. :)

yansern commented:
Got the hosts file working. Safari renders fine. Good to go.

![image](https://user-images.githubusercontent.com/1287077/103540724-79c7cf00-4e9a-11eb-89b7-7880fd8247fb.png)
